src/error.rs

```diff
@@ -0,0 +1,110 @@
+use thiserror::Error;
+
+use crate::spec::ElementType;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// An error that can occur in the `bson` crate.
+#[derive(Debug, Error)]
+#[error("Kind: {kind}")]
+#[non_exhaustive]
+pub struct Error {
+    /// The kind of error that occurred.
+    pub kind: ErrorKind,
+}
+
+/// The types of errors that can occur in the `bson` crate.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// An error occurred when attempting to access a value in a document.
+    #[error("An error occurred when attempting to access a document value for key {key}: {kind}")]
+    #[non_exhaustive]
+    ValueAccess {
+        /// The key of the value.
+        key: String,
```

Comment:
I wonder if it's worth hoisting this to be an optional element of `Error`; I suspect key is relevant data for more types of failure than just this.  Probably this'll become clear as the conversion continues :)